cmd/algorand-indexer/daemon.go

```diff
@@ -65,276 +67,338 @@ var (
+	}
+	return err
+}
 
-		if paramConfigFound {
-			if suppliedAPIConfigFile != "" {
-				logger.Errorf(
-					"api parameter configuration was found in data directory (%s) as well as supplied via command line.  Only provide one.",
-					filepath.Join(indexerDataDir, autoLoadParameterConfigName))
-				panic(exit{1})
-			}
-			suppliedAPIConfigFile = filepath.Join(indexerDataDir, autoLoadParameterConfigName)
-			fmt.Printf("Auto-loading parameter configuration file: %s", suppliedAPIConfigFile)
+func createIndexerPidFile(cfg *daemonConfig) error {
+	var err error
+	logger.Infof("Creating PID file at: %s\n", cfg.pidFilePath)
+	fout, err := os.Create(cfg.pidFilePath)
+	if err != nil {
+		err = fmt.Errorf("%s: could not create pid file, %v", cfg.pidFilePath, err)
+		logger.WithError(err).Error(err)
+		return err
+	}
+	_, err = fmt.Fprintf(fout, "%d", os.Getpid())
+	if err != nil {
+		err = fmt.Errorf("%s: could not write pid file, %v", cfg.pidFilePath, err)
+		logger.WithError(err).Error(err)
+		return err
+	}
+	err = fout.Close()
+	if err != nil {
+		err = fmt.Errorf("%s: could not close pid file, %v", cfg.pidFilePath, err)
+		logger.WithError(err).Error(err)
+		return err
+	}
+	return err
+}
 
-		}
+func runDaemon(daemonConfig *daemonConfig) error {
+	var err error
+	config.BindFlagSet(daemonConfig.flags)
+	err = configureLogger()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to configure logger: %v", err)
+		return err
+	}
 
-		if pidFilePath != "" {
-			fmt.Printf("Creating PID file at: %s\n", pidFilePath)
-			fout, err := os.Create(pidFilePath)
-			maybeFail(err, "%s: could not create pid file, %v", pidFilePath, err)
-			_, err = fmt.Fprintf(fout, "%d", os.Getpid())
-			maybeFail(err, "%s: could not write pid file, %v", pidFilePath, err)
-			err = fout.Close()
-			maybeFail(err, "%s: could not close pid file, %v", pidFilePath, err)
-			defer func(name string) {
-				err := os.Remove(name)
-				if err != nil {
-					logger.WithError(err).Errorf("%s: could not remove pid file", pidFilePath)
-				}
-			}(pidFilePath)
-		}
+	// Create the data directory if necessary/possible
+	if err = configureIndexerDataDir(daemonConfig); err != nil {
+		return err
+	}
 
-		if cpuProfile != "" {
-			var err error
-			profFile, err = os.Create(cpuProfile)
-			maybeFail(err, "%s: create, %v", cpuProfile, err)
-			defer profFile.Close()
-			err = pprof.StartCPUProfile(profFile)
-			maybeFail(err, "%s: start pprof, %v", cpuProfile, err)
-			defer pprof.StopCPUProfile()
-		}
+	// Detect the various auto-loading configs from data directory
+	if err = loadIndexerConfig(daemonConfig); err != nil {
+		return err
+	}
 
-		if configFile != "" {
-			configs, err := os.Open(configFile)
-			if err != nil {
-				maybeFail(err, "%v", err)
-			}
-			defer configs.Close()
-			err = viper.ReadConfig(configs)
+	// Load the Parameter config
+	if err = loadIndexerParamConfig(daemonConfig); err != nil {
+		return err
+	}
+
+	if daemonConfig.pidFilePath != "" {
```

Comment:
You can return a closer function from `createIndexerPidFile` to move more of this logic into the helper function.
```
if daemonConfig.pidFilePath != "" {
    closer, err := createIndexerPidFile(daemonConfig.pidFilePath)
    if err != nil {
        return err
    }
    defer closer()
}
```

Just having fun now... I think something extra fancy could work (but now err has to be defined):
```
var closer func()
var err error
if closer, err = createIndexerPidFile(daemonConfig.pidFilePath); err != nil {
    return err
}
defer closer()
```